include/boost/geometry/algorithms/detail/overlay/get_turn_info_helpers.hpp

```diff
@@ -36,183 +37,309 @@ struct turn_operation_linear
+    inline int qk_wrt_q1() const { return m_side_strategy.apply(get_qi(), get_qj(), get_qk()); }
+
+    inline int pk_wrt_q2() const { return m_side_strategy.apply(get_qj(), get_qk(), get_pk()); }
+    inline int qk_wrt_p2() const { return m_side_strategy.apply(get_pj(), get_pk(), get_qk()); }
+
+    inline point1_type const& get_pi() const { return m_range_p.at(0); }
+    inline point1_type const& get_pj() const { return m_range_p.at(1); }
+    inline point1_type const& get_pk() const { return m_range_p.at(2); }
+
+    inline point2_type const& get_qi() const { return m_range_q.at(0); }
+    inline point2_type const& get_qj() const { return m_range_q.at(1); }
+    inline point2_type const& get_qk() const { return m_range_q.at(2); }
+
+    SideStrategy m_side_strategy; // NOTE: cannot be const&
+    UniqueSubRange1 const& m_range_p;
+    UniqueSubRange2 const& m_range_q;
+};
+
+template<typename Point, typename UniqueSubRange, typename RobustPolicy>
+struct robust_subrange_adapter
+{
+    typedef Point point_type;
+
+    robust_subrange_adapter(UniqueSubRange const& unique_sub_range,
+                     Point const& robust_point_i, Point const& robust_point_j,
+                     RobustPolicy const& robust_policy)
+
+        : m_unique_sub_range(unique_sub_range)
+        , m_robust_policy(robust_policy)
+        , m_robust_point_i(robust_point_i)
+        , m_robust_point_j(robust_point_j)
+        , m_k_retrieved(false)
     {}
 
-    inline int pk_wrt_p1() const { return m_side_strategy.apply(m_pi, m_pj, m_pk); }
-    inline int pk_wrt_q1() const { return m_side_strategy.apply(m_qi, m_qj, m_pk); }
-    inline int qk_wrt_p1() const { return m_side_strategy.apply(m_pi, m_pj, m_qk); }
-    inline int qk_wrt_q1() const { return m_side_strategy.apply(m_qi, m_qj, m_qk); }
+    std::size_t size() const { return m_unique_sub_range.size(); }
+
+    //! Get precalculated point
+    Point const& at(std::size_t index) const
+    {
+        switch (index)
+        {
+            case 0 : return m_robust_point_i;
+            case 1 : return m_robust_point_j;
+            case 2 : return get_point_k();
+            default : return m_robust_point_i;
+        }
+    }
+
+private :
+    Point const& get_point_k() const
+    {
+        if (! m_k_retrieved)
+        {
+            geometry::recalculate(m_robust_point_k, m_unique_sub_range.at(2), m_robust_policy);
+            m_k_retrieved = true;
+        }
+        return m_robust_point_k;
+    }
 
-    inline int pk_wrt_q2() const { return m_side_strategy.apply(m_qj, m_qk, m_pk); }
-    inline int qk_wrt_p2() const { return m_side_strategy.apply(m_pj, m_pk, m_qk); }
+    UniqueSubRange const& m_unique_sub_range;
+    RobustPolicy const& m_robust_policy;
 
-    Pi const& m_pi;
-    Pj const& m_pj;
-    Pk const& m_pk;
-    Qi const& m_qi;
-    Qj const& m_qj;
-    Qk const& m_qk;
+    Point const& m_robust_point_i;
+    Point const& m_robust_point_j;
```

Comment:
So here robust points are stored as references, are we sure that whatever is passed to the constructor of `robust_subrange_adapter` will live longer than an object of this class?